qiskit/transpiler/passes/mapping/stochastic_mapper.py

```diff
@@ -0,0 +1,496 @@
+
+        # If this is the first layer with multi-qubit gates,
+        # output all layers up to this point and ignore any
+        # swap gates. Set the initial layout.
+        if first_layer:
+            logger.debug("layer_update: first multi-qubit gate layer")
+            # Output all layers up to this point
+            for j in range(i + 1):
+                dagcircuit_output.compose_back(layer_list[j]["graph"], layout)
+        # Otherwise, we output the current layer and the associated swap gates.
+        else:
+            # Output any swaps
+            if best_d > 0:
+                logger.debug("layer_update: swaps in this layer, "
+                             "depth %d", best_d)
+                dagcircuit_output.compose_back(best_circ, identity_wire_map)
+            else:
+                logger.debug("layer_update: no swaps in this layer")
+            # Output this layer
+            dagcircuit_output.compose_back(layer_list[i]["graph"], layout)
+
+        return dagcircuit_output
+
+    def _mapper(self, circuit_graph, coupling_graph,
+                initial_layout=None, trials=20, seed=None):
+        """Map a DAGCircuit onto a CouplingGraph using swap gates.
+
+        Args:
+            circuit_graph (DAGCircuit): input DAG circuit
+            coupling_graph (CouplingGraph): coupling graph to map onto
+            initial_layout (dict): dict {(str, int): (str, int)}
+                from qubits of circuit_graph to qubits of
+                coupling_graph (optional)
+            trials (int): number of trials.
+            seed (int): initial seed.
+
+        Returns:
+            DAGCircuit: object containing a circuit equivalent to
+                circuit_graph that respects couplings in coupling_graph
+                and a layout dict mapping qubits of circuit_graph into
+                qubits of coupling_graph. The layout may differ from the
+                initial_layout if the first layer of gates cannot be
+                executed on the initial_layout.
+                Finally, returned is the final layer qubit permutation
+                that is needed to add measurements back in.
+
+        Raises:
+            TranspilerError: if there was any error during the mapping
+                or with the parameters.
+        """
+        if circuit_graph.width() > coupling_graph.size():
+            raise TranspilerError("Not enough qubits in CouplingGraph")
+
+        # Schedule the input circuit
+        layerlist = list(circuit_graph.layers())
+        logger.debug("schedule:")
+        for i, v in enumerate(layerlist):
+            logger.debug("    %d: %s", i, v["partition"])
+
+        if initial_layout is not None:
+            # update initial_layout from a user given
+            # dict{(regname,idx): (regname,idx)}
+            # to an expected dict{(reg,idx): (reg,idx)}
+            device_register = QuantumRegister(coupling_graph.size(), 'q')
```

Comment:
I would like to rework this branch (we have `Layout` object now), but no test takes it. Which input can exercise this code?